fix: never send local endpoints if disabled

deansheather opened the pull request:
Endpoints gathered by scanning interfaces on the local computer (i.e. local addresses like `192.168.1.x`) would get sent to peers via Tailscale Disco protocol, even if the server or client forbade sending endpoints with the server/client flags. This prevents sending any endpoints when blocked by the user or admin.

Previously, we only blocked sending endpoints in the coordination protocol and we blocked STUN servers from functioning, but these local IPs could still get sent via Disco.

Closes #10791

deansheather committed:
fix: never send local endpoints if disabled

kylecarbs reviewed (commented):
Is it possible for us to write a test for this?

deansheather commented:
I can write a test but I don't think it will reliably catch issues. There's no way to force endpoints to update without calling private tailscale methods, so the only way is to sleep and make sure they never appear.

I added multiple thorough tests for the underlying `(*magicsock.Conn).SetBlockEndpoints()` method when I implemented it, which use hidden APIs to settle things quicker.

deansheather committed:
Add 10s delay test to ensure endpoints don't appear

spikecurtis commented on tailnet/conn_test.go:
```diff
@@ -412,9 +413,78 @@ parentLoop:
+			require.Empty(t, node.Endpoints, "conn1 got endpoints")
+		case node := <-nodes2:
+			require.Empty(t, node.Endpoints, "conn2 got endpoints")
+		case <-time.After(testutil.WaitShort):
```
timer gets reset each time we go around the loop, so we could wait for much longer than 10 sec.

But, I don't think we add much value by waiting for 10s.  There are unit tests for `tailnet.Conn` blocking endpoints on node updates.  

What wasn't covered prior to this is endpoints discovered via disco.  The status tests you have below cover that.

deansheather replied:
Oops, yeah I guess I can just remove this section then. I do still think we should wait because we can't force it to gather local endpoints, unless you know a way to test this reliably?

spikecurtis replied:
we have tests that demonstrate that `tailnet.Conn` won't send endpoint in a Node update callback when blockendpoints is set.

E.g. `TestNodeUpdater_setBlockEndpoints_different`

Where we got the endpoints from (local, STUN, etc) is immaterial.

deansheather replied:
I removed the node endpoint check but I still think the delay is required unless you know of another way to force disco endpoints to be gathered and sent immediately

deansheather committed:
Merge branch 'main' into dean/block-endpoints

deansheather committed:
Comments
